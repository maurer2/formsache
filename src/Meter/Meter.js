import React from 'react';
import style from './Meter.module.scss';

const Meter = ({ min, max, value }) => {
  const translations = {
    1: 'bad',
    2: 'middle',
    3: 'good',
  };

  const showDescription = Object.prototype.hasOwnProperty.call(translations, value);

  return (
    <div className={style.wrapper}>
      <meter className={style.meter} min={min} max={max} value={value} />
      { showDescription && (
        <dl className={style.description}>
          <dt className={style.key}>Strength: </dt>
          <dd className={style.value}>{translations[value]}</dd>
        </dl>
      )}
    </div>
  );
};

export default Meter;
