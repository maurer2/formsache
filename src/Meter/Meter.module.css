.wrapper {
  border: 1px solid black;
}

.meter {
  margin: auto;
  width: 50%;
  height: 1rem;
}

.meter::-webkit-meter-bar {
  background: none;
  background-color: none;
  background-image: linear-gradient(
    to right, 
    red 0%,
    red 32.667%,
    transparent 32.667%,
    transparent 33.667%,
    red 33%,
    red 65.337%,
    transparent 65.337%,
    transparent 66.337%,
    red 66.337%,
    red 100%
  );
  background-size: 100% 100%;
}
