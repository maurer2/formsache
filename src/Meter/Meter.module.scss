@function calculate-gradient($total, $color-bar, $color-spacing) {
  $gradients: ();
  $length: percentage(1 / $total);

  @for $i from 1 to ($total + 1) {
    $is-not-last-iteration: ($i != $total);
    $start: ($i - 1) * $length;
    $end: $i * $length;

    $bar: $color-bar $start, $color-bar calc(#{$end} - #{$step-spacing});
    $spacing: $color-spacing calc(#{$end} - #{$step-spacing}), $color-spacing $end;

    $gradient: $bar;

    // last entry doesn't have spacing right
    @if $is-not-last-iteration {
      $gradient: join($bar, $spacing);
    }

    $gradients: join($gradients, $gradient, comma);
  }

  @return $gradients;
}

$meter-background: #fff;
$step-background: lightgray;
$step-foreground: red;
$step-spacing: 10px;

.wrapper {
  border: 1px solid black;
}

.meter {
  position: relative;
  margin: auto;
  width: 50%;
  height: 1rem;
  background: $meter-background;
  background: $step-background;

  // gutter overlay
  &:after {
    $gradients: calculate-gradient(3, transparent, $meter-background);

    content: '';
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    background-image: linear-gradient(to right, $gradients);
    z-index: 55;
  }
}

// background
.meter::-webkit-meter-bar {
  $gradients: calculate-gradient(3, $step-background, $meter-background);

  background: none;
  background-color: inherit;
  background-image: linear-gradient(to right, $gradients);
  background-size: 100%;
  background-repeat: no-repeat;
  box-shadow: none;
}

// actual-bar
.meter::-moz-meter-bar {
  background: none;
  background-color: $step-foreground;
  box-shadow: none;
}

// actual bar
.meter::-webkit-meter-optimum-value,
.meter::-webkit-meter-optimum-value,
.meter::-webkit-meter-suboptimum-value,
.meter::-webkit-meter-even-less-good-value {
  background: none;
  background-color: $step-foreground;
}

.description {
  display: flex;
}

.key {
  font-weight: bold;
}

.value {
  margin: 0;
}
