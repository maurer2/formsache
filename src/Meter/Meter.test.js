import React from 'react';
import { shallow } from 'enzyme';
import Meter from './Meter';

describe('Meter', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Meter
      min={0}
      max={3}
      value={2}
    />);
  });

  test('snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
