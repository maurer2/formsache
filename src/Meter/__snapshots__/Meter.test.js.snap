// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meter snapshot 1`] = `
<Fragment>
  <meter
    className="meter"
    max={3}
    min={0}
    value={2}
  />
  <dl
    className="description"
  >
    <dt
      className="key"
    >
      Strength:
    </dt>
    <dd
      className="value"
    >
      average
    </dd>
  </dl>
</Fragment>
`;
